TombEngine/Specific/Worker.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+namespace TEN::Utils
+{
+	using WorkerTask	  = std::function<void()>;
+	using WorkerTaskGroup = std::vector<WorkerTask>;
+
+	class WorkerManager
+	{
+	private:
+		// Fields
+
+		std::vector<std::thread> _threads	   = {};
+		std::queue<WorkerTask>	 _tasks		   = {};
+		std::mutex				 _taskMutex	   = {};
+		std::condition_variable	 _taskCond	   = {};
+		std::atomic<bool>		 _deinitialize = false;
+
+		std::unordered_map<uint64_t, unsigned int> _groupTaskCounts = {}; // Key = group ID, value = task count.
+		std::atomic<uint64_t>					   _groupIdCounter	= {};
+		std::mutex								   _groupMutex		= {};
+		std::condition_variable					   _groupCond		= {};
+
+	public:
+		// Constructors and destructors
+
+		WorkerManager();
+		~WorkerManager();
+
+		// Getters
+
+		uint64_t	 GetNewGroupId();
+		unsigned int GetThreadCount() const;
+		unsigned int GetCoreCount() const;
+
+		// Utilities
+
+		void AddTask(const WorkerTask& task, uint64_t groupId = (uint64_t)NO_VALUE);
+		void WaitForGroup(uint64_t groupId);
+		
+		// Template for batching parallel operations on vector.
+		template <typename T>
+		void ProcessInParallel(std::vector<T>& vec, const std::function<void(int, int)>& task, bool multiThreaded)
```

Comment:
Consider returning a future to void instead of void so callers can wait on the future when they need to, instead of always having to wait for the completion of the group of tasks. This will further improve parallelism

You may also use a template template to have operations on any container, with your function taking a pair of iterators instead of a pair of indices (also, you should indicate as precondition that task should not change the underlying structure of the vector, otherwise the result of parallel execution is not the same as sequential execution)